core/src/main/java/org/elasticsearch/index/query/GeoPolygonQueryBuilder.java

```diff
@@ -67,39 +90,156 @@ public GeoPolygonQueryBuilder addPoint(GeoPoint point) {
+        // validation was not available prior to 2.x, so to support bwc
+        // percolation queries we only ignore_malformed on 2.x created indexes
+        if (!indexCreatedBeforeV2_0 && !ignoreMalformed) {
+            for (GeoPoint point : shell) {
+                if (point.lat() > 90.0 || point.lat() < -90.0) {
+                    throw new QueryShardException(context, "illegal latitude value [{}] for [{}]", point.lat(),
+                            GeoPolygonQueryBuilder.NAME);
+                }
+                if (point.lon() > 180.0 || point.lon() < -180) {
+                    throw new QueryShardException(context, "illegal longitude value [{}] for [{}]", point.lon(),
+                            GeoPolygonQueryBuilder.NAME);
+                }
+            }
+        }
+
+        if (coerce) {
+            for (GeoPoint point : shell) {
+                GeoUtils.normalizePoint(point, coerce, coerce);
+            }
+        }
+
+        MappedFieldType fieldType = context.fieldMapper(fieldName);
+        if (fieldType == null) {
+            throw new QueryShardException(context, "failed to find geo_point field [" + fieldName + "]");
+        }
+        if (!(fieldType instanceof GeoPointFieldMapper.GeoPointFieldType)) {
+            throw new QueryShardException(context, "field [" + fieldName + "] is not a geo_point field");
+        }
+
+        IndexGeoPointFieldData indexFieldData = context.getForField(fieldType);
+        return new GeoPolygonQuery(indexFieldData, shell.toArray(new GeoPoint[shell.size()]));
+    }
+
     @Override
     protected void doXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startObject(NAME);
 
-        builder.startObject(name);
-        builder.startArray(POINTS);
+        builder.startObject(fieldName);
+        builder.startArray(GeoPolygonQueryParser.POINTS_FIELD.getPreferredName());
         for (GeoPoint point : shell) {
             builder.startArray().value(point.lon()).value(point.lat()).endArray();
         }
         builder.endArray();
         builder.endObject();
 
-        if (coerce != null) {
-            builder.field("coerce", coerce);
-        }
-        if (ignoreMalformed != null) {
-            builder.field("ignore_malformed", ignoreMalformed);
-        }
+        builder.field(GeoPolygonQueryParser.COERCE_FIELD.getPreferredName(), coerce);
+        builder.field(GeoPolygonQueryParser.IGNORE_MALFORMED_FIELD.getPreferredName(), ignoreMalformed);
+
         printBoostAndQueryName(builder);
         builder.endObject();
     }
 
     @Override
+    protected GeoPolygonQueryBuilder doReadFrom(StreamInput in) throws IOException {
+        String fieldName = in.readString();
+        List<GeoPoint> shell = new ArrayList<>();
+        int size = in.readVInt();
+        for (int i = 0; i < size; i++) {
+            shell.add(new GeoPoint(in.readDouble(), in.readDouble()));
```

Comment:
IMHO we can also postpone such clean-up until after the branch is merged to master, just need to remember that we left a few things behind.